python/ray/_private/node.py

```diff
@@ -964,6 +972,120 @@ def _get_cached_port(
+        Otherwise, the file exists, but it is emptyl.
+
+        This API is process-safe, meaning the file access is protected by
+        a file lock.
+
+        Returns:
+            node_ip_address cached on the current node. None if the node
+            the file doesn't exist, meaning ray instance hasn't been
+            started on a current node. If node_ip_address is not written
+            to a file, it means --node-ip-address is not given, and in this
+            case, we find the IP address ourselves.
+        """
+        assert hasattr(self, "_session_dir")
+        file_path = Path(
+            os.path.join(self.get_session_dir_path(), "node_ip_address.json")
+        )
+        cached_node_ip_address = {}
+
+        with FileLock(str(file_path.absolute()) + ".lock"):
+            if not file_path.exists():
+                return None
+
+            with file_path.open() as f:
+                cached_node_ip_address.update(json.load(f))
+
+            if "node_ip_address" in cached_node_ip_address:
+                return cached_node_ip_address["node_ip_address"]
+            else:
+                return ray.util.get_node_ip_address()
+
+    def _write_node_ip_address(self, node_ip_address: Optional[str]) -> None:
+        """Write a node ip address of the current session to
+        node_ip_address.json.
+
+        If a ray instance is started by `ray start --node-ip-address`,
+        the node ip address is cached to a file node_ip_address.json.
+
+        This API is process-safe, meaning the file access is protected by
+        a file lock.
+
+        The file contains a single string node_ip_address. It nothing
+        is written, --node-ip-address is not given. In this case, Ray
+        resolves the IP address on its own. It assumes in a single node,
+        you can have only 1 IP address (which is the assumption ray
+        has in general).
+
+        node_ip_address is the ip address of the current node.
+
+        Args:
+            node_ip_address: The node IP address of the current node.
+                If None, it means the node ip address is not given
+                by --node-ip-address. In this case, we don't write
+                anything to a file.
+        """
+        assert hasattr(self, "_session_dir")
+
+        file_path = Path(
+            os.path.join(self.get_session_dir_path(), "node_ip_address.json")
+        )
+        cached_node_ip_address = {}
+
+        with FileLock(str(file_path.absolute()) + ".lock"):
+            if not file_path.exists():
+                with file_path.open(mode="w") as f:
+                    json.dump({}, f)
+
+            with file_path.open() as f:
+                cached_node_ip_address.update(json.load(f))
+
+            if node_ip_address is not None:
+                cached_node_ip_address["node_ip_address"] = node_ip_address
+                with file_path.open(mode="w") as f:
+                    json.dump(cached_node_ip_address, f)
```

Comment:
This part seems weird. What if, file path exists and it's not empty?

I'm not sure whether this is right or not:

- check whether the file exists, if yes, load it compare it with the node ip address provided. error or overwrite
- if it doesn't exist, check the node ip address provided and write either empty or the node ip address json.

Reply:
I think the assumption is it won't have scenario where the contents in the file is changed. This can only happen if we start 2 ray instances using different ip addresses. It'd be really difficult to support this case (when we don't know the IP address, we can't get the Node ID. But if we can have 2 instances with different IDs, we need some sort of IDs. It is a circular dependency problem). 

But I understand your concern, and I made it this way;
1. If content is already written, and if the IP is the same, no-op.
2. If IP is different, we will write a warning that the IP will be overwritten (I can also consider assert, but it may be a little risky to add assert here in the last minute cherry pick)